src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -3065,6 +3064,23 @@ public void report(DiagnosticPosition _unused, JCDiagnostic details) {
             };
         }
 
+        void checkNewInnerClass(DiagnosticPosition pos, Env<AttrContext> env, Type type) {
+            boolean isLocal = type.tsym.owner.kind == MTH;
```

Comment:
This is a bit subtle - we need to run the new checks for _all_ inner classes. E.g. if a local class `C` has no enclosing instance (because is declared in a static context), we might still need to check that `new C()` is valid, because `C` might capture variables from the enclosing (static) method.